<!DOCTYPE html>
<html lang="en">
<head>
    <title>
Error Management: Future vs ZIO - linksfor.dev(s)    </title>
    <meta charset="utf-8">
    <link rel="alternate" type="application/rss+xml" title="Linksfor.dev(s) feed" href="https://linksfor.dev/feed.rss" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <meta name="google" value="notranslate">
    <link rel="stylesheet" type="text/css" href="/style.css" />
    <link rel="shortcut icon" href="/favicon.ico" type="image/x-icon">
    <link rel="icon" href="/favicon.ico" type="image/x-icon">
    <meta property="og:title" content="Error Management: Future vs ZIO - linksfor.dev(s)"/>
    <meta property="article:author" content="John De Goes"/>
    <meta property="og:description" content="Modern applications are concurrent, parallel, asynchronous, and synchronous; they utilize many different subsystems, including network systems, actor systems, &#x2026;"/>
    <meta property="og:type" content="website"/>
    <meta property="og:url" content="https://www.slideshare.net/jdegoes/error-management-future-vs-zio"/>

<meta property="og:site_name" content="linksfor.dev(s)" />
</head>
<body>
    <div class="grid">
        <h1>
                <span style="cursor: default" title="linksfor.dev(s) has been running for 1 year! :partypopper:">üéâ</span>
<a href="/" style="color:inherit">linksfor.dev(s)</a>
        </h1>
        <title>linksfor.dev(s) - Error Management: Future vs ZIO</title>
<div class="readable">
        <h1>Error Management: Future vs ZIO</h1>
        <p>
by John De Goes <br/>Reading time: 18-23 minutes        </p>
        <p><a href="https://www.slideshare.net/jdegoes/error-management-future-vs-zio">https://www.slideshare.net/jdegoes/error-management-future-vs-zio</a></p>
        <hr/>
<div id="readability-page-1" class="page"><div>
        
        


      

        <p><span>Successfully reported this slideshow.</span>
  </p>


      
















<div id="slideview-container">
  

  <div>

    <div id="main-panel">


      

      <div itemscope="" itemtype="https://schema.org/MediaObject">


        <div>
          <meta itemprop="inLanguage" content="en">
          <meta itemprop="image" content="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-1-638.jpg?cb=1557619013">
          <meta itemprop="thumbnailUrl" content="https://cdn.slidesharecdn.com/ss_thumbnails/errormanagementfuturevszio-190511235539-thumbnail.jpg?cb=1557619013">
          <meta itemprop="embedURL" content="https://www.slideshare.net/slideshow/embed_code/key/tPnd4O9WhlXLAB">
          <meta itemprop="playerType" content="HTML5">
          <meta itemprop="interactionCount" content="UserComments:5">
          <meta itemprop="interactionCount" content="UserLikes:20">
          <meta itemprop="interactionCount" content="UserDownloads:75">
          <meta itemprop="interactionCount" content="UserPageVisits:8844">
          <meta itemprop="interactionCount" content="UserPlays:8844">
          <meta itemprop="interactionCount" content="UserPlusOnes:0" id="meta-google">


          

          <ul id="slideshow-actions">
            <li>
              
            </li>
              <li>
                
              </li>
              <li>
                
              </li>

            <!-- Report/Flag Content -->
            <li>
              <a data-dropdown="report-dropdown" aria-controls="report-dropdown" aria-expanded="false">
                <span>...</span></a>
              <span title="You have reported this as inappropriate">
                <i></i>
                <span aria-label="You have reported this as inappropriate"></span>
              </span>
              
            </li>
          </ul>


          <div itemprop="author" itemscope="" itemtype="https://schema.org/Person">

            <p><a href="https://www.slideshare.net/jdegoes?utm_campaign=profiletracking&amp;utm_medium=sssite&amp;utm_source=ssslideview" title="jdegoes" itemprop="url">
                <img alt="John De Goes" itemprop="image" src="https://cdn.slidesharecdn.com/profile-photo-jdegoes-48x48.jpg?cb=1565359801">
              </a>
            </p>

            
          </div>

            

          <p>
            <small>
              Published on <time datetime="May 11, 2019" itemprop="datePublished">May 11, 2019</time>
            </small>
          </p>

            <div>
              <div data-ga-cat="bigfoot_slideview" data-ga-action="description>more">
                <p id="slideshow-description-paragraph">
                    Modern applications are concurrent, parallel, asynchronous, and synchronous; </p>
                
              </div>
            </div>


            



        </div>


        

            <div>
              
              <ol itemprop="text">
                    <li>
      1.
    1
Error Management
ZIO vs Future
Dublin Scala Meetup, May 11th
John A. De Goes &nbsp; @jdegoes
Kai @kaidaxofÔ¨Åcial
...with help of Pavel Shirshov @shirshovp
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-2-638.jpg?cb=1557619013" title="Agenda
A Tale of Two Effects
2
Next-Gen Debugging Conclusio..." target="_blank">
        2.
      </a>
    Agenda
A Tale of Two Effects
2
Next-Gen Debugging Conclusion
Managing Errors
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-3-638.jpg?cb=1557619013" title="3
A TALE OF
TWO EFFECTS
FUTURE
 " target="_blank">
        3.
      </a>
    3
A TALE OF
TWO EFFECTS
FUTURE
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-4-638.jpg?cb=1557619013" title="Parallel
Future enables parallel
computations and
non-block..." target="_blank">
        4.
      </a>
    Parallel
Future enables parallel
computations and
non-blocking gathering.
Error
Future has a built-in error
channel for
Throwable-based errors.
Eager
Future is not referentially
transparent; refactoring
may change behavior.
4
FUTURE
Async
Future enables
non-blocking code that
eÔ¨Éciently uses threads.
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-5-638.jpg?cb=1557619013" title="5
Future ZIO Effect
go running
 " target="_blank">
        5.
      </a>
    5
Future ZIO Effect
go running
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-6-638.jpg?cb=1557619013" title="Procedural Functional
def program: Unit = {
println(&quot;What‚Äôs..." target="_blank">
        6.
      </a>
    Procedural Functional
def program: Unit = {
println("What‚Äôs your name?")
val name = readLine()
println(s"Howdy. $name!")
return ()
}
val program = for {
_ &lt;- putStrLn("What‚Äôs your name?")
name &lt;- getStrLn
_ &lt;- putStrLn(s"Howdy, $name")
} yield ()
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-7-638.jpg?cb=1557619013" title="7
Future ZIO
Performance > 100x slower > 100x faster
Cancel..." target="_blank">
        7.
      </a>
    7
Future ZIO
Performance &gt; 100x slower &gt; 100x faster
Cancellation &amp; Timeouts êÑÇ ‚úî
Effect Combinators êÑÇ ‚úî
Resource Safety êÑÇ ‚úî
Fiber Concurrency êÑÇ ‚úî
Equational &amp; Type Reasoning êÑÇ ‚úî
Testability êÑÇ ‚úî
Error Management &amp; Debugging ? ?
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-8-638.jpg?cb=1557619013" title="8
ZIO[R, E, A]
Environment Type
Failure Type
Success Type
 " target="_blank">
        8.
      </a>
    8
ZIO[R, E, A]
Environment Type
Failure Type
Success Type
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-9-638.jpg?cb=1557619013" title="9
ZIO[R, E, A]
~
R => Either[E, A]
 " target="_blank">
        9.
      </a>
    9
ZIO[R, E, A]
~
R =&gt; Either[E, A]
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-10-638.jpg?cb=1557619013" title="10
type Task [ +A] = ZIO[Any, Throwable, A]
type UIO [ +A] ..." target="_blank">
        10.
      </a>
    10
type Task [ +A] = ZIO[Any, Throwable, A]
type UIO [ +A] = ZIO[Any, Nothing, A]
type TaskR[+R,+A] = ZIO[ R, Throwable, A]
type IO [+E,+A] = ZIO[Any, E, A]
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-11-638.jpg?cb=1557619013" title="11
type Task [ +A] = ZIO[Any, Throwable, A]
type UIO [ +A] ..." target="_blank">
        11.
      </a>
    11
type Task [ +A] = ZIO[Any, Throwable, A]
type UIO [ +A] = ZIO[Any, Nothing, A]
type TaskR[+R,+A] = ZIO[ R, Throwable, A]
type IO [+E,+A] = ZIO[Any, E, A]
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-12-638.jpg?cb=1557619013" title="12
type Task [ +A] = ZIO[Any, Throwable, A]
type UIO [ +A] ..." target="_blank">
        12.
      </a>
    12
type Task [ +A] = ZIO[Any, Throwable, A]
type UIO [ +A] = ZIO[Any, Nothing, A]
type TaskR[+R,+A] = ZIO[ R, Throwable, A]
type IO [+E,+A] = ZIO[Any, E, A]
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-13-638.jpg?cb=1557619013" title="13
type Task [ +A] = ZIO[Any, Throwable, A]
type UIO [ +A] ..." target="_blank">
        13.
      </a>
    13
type Task [ +A] = ZIO[Any, Throwable, A]
type UIO [ +A] = ZIO[Any, Nothing, A]
type TaskR[+R,+A] = ZIO[ R, Throwable, A]
type IO [+E,+A] = ZIO[Any, E, A]
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-14-638.jpg?cb=1557619013" title="ZIO[R, E, A]
Synchronous Asynchronous Errors Resource
ZIO.s..." target="_blank">
        14.
      </a>
    ZIO[R, E, A]
Synchronous Asynchronous Errors Resource
ZIO.succeed(‚Ä¶)
ZIO.effect(‚Ä¶)
ZIO.effectTotal(‚Ä¶)
effectBlocking(‚Ä¶)
ZIO.effectAsync(‚Ä¶)
ZIO.effectAsyncMaybe(‚Ä¶)
ZIO.effectAsyncInte‚Ä¶(‚Ä¶)
ZIO.fromFuture(‚Ä¶)
ZIO.fail(‚Ä¶)
ZIO.fromOption(‚Ä¶)
ZIO.fromEither(‚Ä¶)
ZIO.fromTry(‚Ä¶)
ZIO.bracket(‚Ä¶)
ZIO.reserve(‚Ä¶)
ZIO.ensuring(‚Ä¶)
fromAutoCloseable(‚Ä¶)
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-15-638.jpg?cb=1557619013" title="15
ERROR
MANAGEMENT
FUTURE
 " target="_blank">
        15.
      </a>
    15
ERROR
MANAGEMENT
FUTURE
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-16-638.jpg?cb=1557619013" title="Fail Domain errors, business errors,
transient errors, expe..." target="_blank">
        16.
      </a>
    Fail Domain errors, business errors,
transient errors, expected errors...
Expected Errors
Not ReÔ¨Çected in Types
DieSystem errors, fatal errors,
unanticipated errors, defects...
Unexpected Errors
ReÔ¨Çected in Types
16
ERROR DUALITY
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-17-638.jpg?cb=1557619013" title="17
ERROR DUALITY
Not ReÔ¨Çected in TypesNot ReÔ¨Çected in Types..." target="_blank">
        17.
      </a>
    17
ERROR DUALITY
Not ReÔ¨Çected in TypesNot ReÔ¨Çected in Types
val failed: Future[Nothing] =
Future.failed(new Exception)
val died: Future[Nothing] =
Future(throw new Error)
FUTURE
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-18-638.jpg?cb=1557619013" title="18
ERROR DUALITY
Not ReÔ¨Çected in TypesReÔ¨Çected in Types
val..." target="_blank">
        18.
      </a>
    18
ERROR DUALITY
Not ReÔ¨Çected in TypesReÔ¨Çected in Types
val failed: IO[String, Nothing] =
ZIO.fail(‚ÄúUh oh!‚Äù)
val died: IO[Nothing, Nothing] =
ZIO.dieMessage(‚ÄúUh oh!‚Äù)
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-19-638.jpg?cb=1557619013" title="19
ERROR COMPOSITION
e1
e2t
First Error
 " target="_blank">
        19.
      </a>
    19
ERROR COMPOSITION
e1
e2t
First Error
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-20-638.jpg?cb=1557619013" title="20
ERROR COMPOSITION
e1
e2t
Second Error
 " target="_blank">
        20.
      </a>
    20
ERROR COMPOSITION
e1
e2t
Second Error
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-21-638.jpg?cb=1557619013" title="21
ERROR COMPOSITION
e1
e2tFinalizer Error
 " target="_blank">
        21.
      </a>
    21
ERROR COMPOSITION
e1
e2tFinalizer Error
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-22-638.jpg?cb=1557619013" title="22
ERROR COMPOSITION
e1
e2t
?
 " target="_blank">
        22.
      </a>
    22
ERROR COMPOSITION
e1
e2t
?
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-23-638.jpg?cb=1557619013" title="23
ERROR COMPOSITION
e1
e2t
FUTURE
Thrown away!!!Reported o..." target="_blank">
        23.
      </a>
    23
ERROR COMPOSITION
e1
e2t
FUTURE
Thrown away!!!Reported on side channel!!!
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-24-638.jpg?cb=1557619013" title="24
ERROR COMPOSITION
e1
e2t
Cause[E]
 " target="_blank">
        24.
      </a>
    24
ERROR COMPOSITION
e1
e2t
Cause[E]
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-25-638.jpg?cb=1557619013" title="25
ERROR COMPOSITION
e1
e2t
Cause[E]
Cause.Fail(e1)
 " target="_blank">
        25.
      </a>
    25
ERROR COMPOSITION
e1
e2t
Cause[E]
Cause.Fail(e1)
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-26-638.jpg?cb=1557619013" title="26
ERROR COMPOSITION
e1
e2t
Cause[E]
Cause.Both(
Cause.Fail..." target="_blank">
        26.
      </a>
    26
ERROR COMPOSITION
e1
e2t
Cause[E]
Cause.Both(
Cause.Fail(e1),
Cause.Fail(e2))
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-27-638.jpg?cb=1557619013" title="27
ERROR COMPOSITION
e1
e2t
Cause[E]
Cause.Both(
Cause.Fail..." target="_blank">
        27.
      </a>
    27
ERROR COMPOSITION
e1
e2t
Cause[E]
Cause.Both(
Cause.Fail(e1),
Cause.Then(
Cause.Fail(e2),
Cause.Die(t))
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-28-638.jpg?cb=1557619013" title="zio.FiberFailure: Fiber failed.
‚ï•
‚ï†‚ïê‚ïê‚ïó
‚ïë ‚ïë
‚ïë ‚ïë
‚ïë ‚ïë
‚ïë ‚ï†‚îÄ A c..." target="_blank">
        28.
      </a>
    zio.FiberFailure: Fiber failed.
‚ï•
‚ï†‚ïê‚ïê‚ïó
‚ïë ‚ïë
‚ïë ‚ïë
‚ïë ‚ïë
‚ïë ‚ï†‚îÄ A checked error was not handled:
‚ïë ‚ïë
Failed(DatabaseUnreachableError)
‚ïë ‚ïë
‚ïë ‚ï†‚îÄ A finalizer threw an error:
‚ïë ‚ñº
Die(IndexOutOfBoundsException())
‚ïë
‚ï†‚îÄ A checked error was not handled:
‚ñº Failed(UserIdNotFoundError) 28
ERROR COMPOSITION
e2
t
e1
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-29-638.jpg?cb=1557619013" title="Asynchronous
29
ERROR PROPAGATION
Synchronous
ConcurrentPar..." target="_blank">
        29.
      </a>
    Asynchronous
29
ERROR PROPAGATION
Synchronous
ConcurrentParallel
Resource
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-30-638.jpg?cb=1557619013" title="Asynchronous
30
ERROR PROPAGATION
Synchronous
ConcurrentPar..." target="_blank">
        30.
      </a>
    Asynchronous
30
ERROR PROPAGATION
Synchronous
ConcurrentParallel
Resource
FUTURE
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-31-638.jpg?cb=1557619013" title="Asynchronous
31
ERROR PROPAGATION
Synchronous
ConcurrentPar..." target="_blank">
        31.
      </a>
    Asynchronous
31
ERROR PROPAGATION
Synchronous
ConcurrentParallel
Resource
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-32-638.jpg?cb=1557619013" title="32
ERROR RECOVERY
Fallback Catching Folding Value
ZIO#orEls..." target="_blank">
        32.
      </a>
    32
ERROR RECOVERY
Fallback Catching Folding Value
ZIO#orElse(‚Ä¶)
ZIO#orElseEither(‚Ä¶)
ZIO#catchAll(‚Ä¶)
ZIO#catchSome(‚Ä¶)
ZIO#fold(‚Ä¶, ‚Ä¶)
ZIO#foldM(‚Ä¶, ‚Ä¶)
effect1.orElse(effect2) future1.fallback(future2)
ZIO#either
ZIO#run
FUTURE
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-33-638.jpg?cb=1557619013" title="33
ERROR RECOVERY
Fallback Catching Folding Value
ZIO#orEls..." target="_blank">
        33.
      </a>
    33
ERROR RECOVERY
Fallback Catching Folding Value
ZIO#orElse(‚Ä¶)
ZIO#orElseEither(‚Ä¶)
ZIO#catchAll(‚Ä¶)
ZIO#catchSome(‚Ä¶)
ZIO#fold(‚Ä¶, ‚Ä¶)
ZIO#foldM(‚Ä¶, ‚Ä¶)
effect.catchAll(f) future.recoverWith(pf)
ZIO#either
ZIO#run
FUTURE
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-34-638.jpg?cb=1557619013" title="34
ERROR RECOVERY
Fallback Catching Folding Value
ZIO#orEls..." target="_blank">
        34.
      </a>
    34
ERROR RECOVERY
Fallback Catching Folding Value
ZIO#orElse(‚Ä¶)
ZIO#orElseEither(‚Ä¶)
ZIO#catchAll(‚Ä¶)
ZIO#catchSome(‚Ä¶)
ZIO#fold(‚Ä¶, ‚Ä¶)
ZIO#foldM(‚Ä¶, ‚Ä¶)
effect.catchSome(pf) future.recoverWith(pf)
ZIO#either
ZIO#run
FUTURE
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-35-638.jpg?cb=1557619013" title="35
ERROR RECOVERY
Fallback Catching Folding Value
ZIO#orEls..." target="_blank">
        35.
      </a>
    35
ERROR RECOVERY
Fallback Catching Folding Value
ZIO#orElse(‚Ä¶)
ZIO#orElseEither(‚Ä¶)
ZIO#catchAll(‚Ä¶)
ZIO#catchSome(‚Ä¶)
ZIO#fold(‚Ä¶, ‚Ä¶)
ZIO#foldM(‚Ä¶, ‚Ä¶)
effect.fold(err, succ)
ZIO#either
ZIO#run
FUTURE
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-36-638.jpg?cb=1557619013" title="36
ERROR RECOVERY
Fallback Catching Folding Value
ZIO#orEls..." target="_blank">
        36.
      </a>
    36
ERROR RECOVERY
Fallback Catching Folding Value
ZIO#orElse(‚Ä¶)
ZIO#orElseEither(‚Ä¶)
ZIO#catchAll(‚Ä¶)
ZIO#catchSome(‚Ä¶)
ZIO#fold(‚Ä¶, ‚Ä¶)
ZIO#foldM(‚Ä¶, ‚Ä¶)
effect.foldM(err, succ)
ZIO#either
ZIO#run
FUTURE
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-37-638.jpg?cb=1557619013" title="37
ERROR RECOVERY
Fallback Catching Folding Value
ZIO#orEls..." target="_blank">
        37.
      </a>
    37
ERROR RECOVERY
Fallback Catching Folding Value
ZIO#orElse(‚Ä¶)
ZIO#orElseEither(‚Ä¶)
ZIO#catchAll(‚Ä¶)
ZIO#catchSome(‚Ä¶)
ZIO#fold(‚Ä¶, ‚Ä¶)
ZIO#foldM(‚Ä¶, ‚Ä¶)
effect.either
ZIO#either
ZIO#run
FUTURE
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-38-638.jpg?cb=1557619013" title="38
ERROR RECOVERY
Fallback Catching Folding Value
ZIO#orEls..." target="_blank">
        38.
      </a>
    38
ERROR RECOVERY
Fallback Catching Folding Value
ZIO#orElse(‚Ä¶)
ZIO#orElseEither(‚Ä¶)
ZIO#catchAll(‚Ä¶)
ZIO#catchSome(‚Ä¶)
ZIO#fold(‚Ä¶, ‚Ä¶)
ZIO#foldM(‚Ä¶, ‚Ä¶)
ZIO#either
ZIO#run
effect.run
FUTURE
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-39-638.jpg?cb=1557619013" title="39
BEST
PRACTICES
 " target="_blank">
        39.
      </a>
    39
BEST
PRACTICES
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-40-638.jpg?cb=1557619013" title="40
1. DON‚ÄôT TYPE UNEXPECTED ERRORS
ZIO.effect(httpClient.ge..." target="_blank">
        40.
      </a>
    40
1. DON‚ÄôT TYPE UNEXPECTED ERRORS
ZIO.effect(httpClient.get(url)).refineOrDie {
case e : TemporarilyUnavailable =&gt; e
}.retry(RetryPolicy).orDie
: IO[TemporarilyUnavailable, Response]
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-41-638.jpg?cb=1557619013" title="41
2. DO EXTEND EXCEPTION WITH SEALED TRAITS
sealed trait U..." target="_blank">
        41.
      </a>
    41
2. DO EXTEND EXCEPTION WITH SEALED TRAITS
sealed trait UserServiceError
extends Exception
case class InvalidUserId(id: ID)
extends UserServiceError
case class ExpiredAuth(id: ID)
extends UserServiceError
UserServiceError
InvalidUserId ExpiredAuth
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-42-638.jpg?cb=1557619013" title="42
2. DO EXTEND EXCEPTION WITH SEALED TRAITS
userServiceErr..." target="_blank">
        42.
      </a>
    42
2. DO EXTEND EXCEPTION WITH SEALED TRAITS
userServiceError match {
case InvalidUserId(id) =&gt; ...
case ExpiredAuth(id) =&gt; ...
}
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-43-638.jpg?cb=1557619013" title="43
2. DO EXTEND EXCEPTION WITH SEALED TRAITS
for {
service ..." target="_blank">
        43.
      </a>
    43
2. DO EXTEND EXCEPTION WITH SEALED TRAITS
for {
service &lt;- userAuth(token)
_ &lt;- service.userProfile(userId)
body &lt;- generateEmail(orderDetails)
receipt &lt;- sendEmail(‚ÄúYour Order Details‚Äù,
body, profile.email)
} yield receipt
ExpiredAuth
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-44-638.jpg?cb=1557619013" title="44
2. DO EXTEND EXCEPTION WITH SEALED TRAITS
for {
service ..." target="_blank">
        44.
      </a>
    44
2. DO EXTEND EXCEPTION WITH SEALED TRAITS
for {
service &lt;- userAuth(token)
_ &lt;- service.userProfile(userId)
body &lt;- generateEmail(orderDetails)
receipt &lt;- sendEmail(‚ÄúYour Order Details‚Äù,
body, profile.email)
} yield receipt
InvalidUserId
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-45-638.jpg?cb=1557619013" title="45
2. DO EXTEND EXCEPTION WITH SEALED TRAITS
for {
service ..." target="_blank">
        45.
      </a>
    45
2. DO EXTEND EXCEPTION WITH SEALED TRAITS
for {
service &lt;- userAuth(token)
_ &lt;- service.userProfile(userId)
body &lt;- generateEmail(orderDetails)
receipt &lt;- sendEmail(‚ÄúYour Order Details‚Äù,
body, profile.email)
} yield receipt
Nothing
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-46-638.jpg?cb=1557619013" title="46
2. DO EXTEND EXCEPTION WITH SEALED TRAITS
for {
service ..." target="_blank">
        46.
      </a>
    46
2. DO EXTEND EXCEPTION WITH SEALED TRAITS
for {
service &lt;- userAuth(token)
_ &lt;- service.userProfile(userId)
body &lt;- generateEmail(orderDetails)
receipt &lt;- sendEmail(‚ÄúYour Order Details‚Äù,
body, profile.email)
} yield receipt
EmailDeliveryError
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-47-638.jpg?cb=1557619013" title="47
2. DO EXTEND EXCEPTION WITH SEALED TRAITS
for {
service ..." target="_blank">
        47.
      </a>
    47
2. DO EXTEND EXCEPTION WITH SEALED TRAITS
for {
service &lt;- userAuth(token)
_ &lt;- service.userProfile(userId)
body &lt;- generateEmail(orderDetails)
receipt &lt;- sendEmail(‚ÄúYour Order Details‚Äù,
body, profile.email)
} yield receipt
IO[Exception, Receipt]
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-48-638.jpg?cb=1557619013" title="48
3. DON‚ÄôT REFLEXIVELY LOG ERRORS
uploadFile(‚Äúcontacts.csv..." target="_blank">
        48.
      </a>
    48
3. DON‚ÄôT REFLEXIVELY LOG ERRORS
uploadFile(‚Äúcontacts.csv‚Äù).catchAll { error =&gt;
// Log error and re-fail:
for {
_ &lt;- logger.error(error)
_ &lt;- ZIO.fail(error)
} yield ()
}
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-49-638.jpg?cb=1557619013" title="49
4. DO GET TO KNOW UIO
type UIO[+A] = ZIO[Any, Nothing, A..." target="_blank">
        49.
      </a>
    49
4. DO GET TO KNOW UIO
type UIO[+A] = ZIO[Any, Nothing, A]
Cannot fail!
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-50-638.jpg?cb=1557619013" title="lazy val processed: UIO[Unit] =
processUpload(upload).eithe..." target="_blank">
        50.
      </a>
    lazy val processed: UIO[Unit] =
processUpload(upload).either.flatMap {
case Left (_) =&gt; processed.delay(1.minute)
case Right(_) =&gt; ZIO.succeed(())
}
50
4. DO GET TO KNOW UIO
Fails with UploadError
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-51-638.jpg?cb=1557619013" title="lazy val processed: UIO[Unit] =
processUpload(upload).eithe..." target="_blank">
        51.
      </a>
    lazy val processed: UIO[Unit] =
processUpload(upload).either.flatMap {
case Left (_) =&gt; processed.delay(1.minute)
case Right(_) =&gt; ZIO.succeed(())
}
51
4. DO GET TO KNOW UIO
Fails with Nothing
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-52-638.jpg?cb=1557619013" title="52
4. DO GET TO KNOW UIO
lazy val processed: UIO[Unit] =
pr..." target="_blank">
        52.
      </a>
    52
4. DO GET TO KNOW UIO
lazy val processed: UIO[Unit] =
processUpload(upload).either.flatMap {
case Left (_) =&gt; processed.delay(1.minute)
case Right(_) =&gt; ZIO.succeed(())
}
Fails with Nothing
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-53-638.jpg?cb=1557619013" title="NEXT-GENERATION
DEBUGGING
 " target="_blank">
        53.
      </a>
    NEXT-GENERATION
DEBUGGING
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-54-638.jpg?cb=1557619013" title="54
def asyncDbCall(sql: SQL): Future[Result]
def selectHuma..." target="_blank">
        54.
      </a>
    54
def asyncDbCall(sql: SQL): Future[Result]
def selectHumans(): Future[Result] = ...asyncDbCall(...)...
def selectPets(): Future[Result] = ...asyncDbCall(...)...
FUTURE
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-55-638.jpg?cb=1557619013" title="55
def asyncDbCall(sql: SQL): Future[Result]
def selectHuma..." target="_blank">
        55.
      </a>
    55
def asyncDbCall(sql: SQL): Future[Result]
def selectHumans(): Future[Result] = ...asyncDbCall(...)...
def selectPets(): Future[Result] = ...asyncDbCall(...)...
FUTURE
Which function failed, selectHumans or selectPets?
PostgresException: Syntax error at or near 42
at example$.getConnection(example.scala:43)
at example$.$anonfun$asyncDbCall$1(example.scala:23)
at scala.concurrent.Future$.$anonfun$apply$1(Future.scala:658)
at scala.util.Success.$anonfun$map$1(Try.scala:255)
at scala.util.Success.map(Try.scala:213)
at scala.concurrent.Future.$anonfun$map$1(Future.scala:292)
at scala.concurrent.impl.Promise.liftedTree1$1(Promise.scala:33)
at scala.concurrent.impl.Promise.$anonfun$transform$1(Promise.scala:33)
at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:64)
at java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1402)
at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-56-638.jpg?cb=1557619013" title="56
Which function failed, selectHumans or selectPets?
FUTUR..." target="_blank">
        56.
      </a>
    56
Which function failed, selectHumans or selectPets?
FUTURE
def asyncDbCall(sql: SQL): Future[Result]
def selectHumans(): Future[Result] = ...asyncDbCall(...)...
def selectPets(): Future[Result] = ...asyncDbCall(...)...
Only the last operation is mentioned
PostgresException: Syntax error at or near 42
at example$.getConnection(example.scala:43)
at example$.$anonfun$asyncDbCall$1(example.scala:23)
at scala.concurrent.Future$.$anonfun$apply$1(Future.scala:658)
at scala.util.Success.$anonfun$map$1(Try.scala:255)
at scala.util.Success.map(Try.scala:213)
at scala.concurrent.Future.$anonfun$map$1(Future.scala:292)
at scala.concurrent.impl.Promise.liftedTree1$1(Promise.scala:33)
at scala.concurrent.impl.Promise.$anonfun$transform$1(Promise.scala:33)
at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:64)
at java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1402)
at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-57-638.jpg?cb=1557619013" title="57
Which function failed, selectHumans or selectPets?
FUTUR..." target="_blank">
        57.
      </a>
    57
Which function failed, selectHumans or selectPets?
FUTURE
Only the last operation is mentioned
There is NO way to know!!!
def asyncDbCall(sql: SQL): Future[Result]
def selectHumans(): Future[Result] = ...asyncDbCall(...)...
def selectPets(): Future[Result] = ...asyncDbCall(...)...
PostgresException: Syntax error at or near 42
at example$.getConnection(example.scala:43)
at example$.$anonfun$asyncDbCall$1(example.scala:23)
at scala.concurrent.Future$.$anonfun$apply$1(Future.scala:658)
at scala.util.Success.$anonfun$map$1(Try.scala:255)
at scala.util.Success.map(Try.scala:213)
at scala.concurrent.Future.$anonfun$map$1(Future.scala:292)
at scala.concurrent.impl.Promise.liftedTree1$1(Promise.scala:33)
at scala.concurrent.impl.Promise.$anonfun$transform$1(Promise.scala:33)
at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:64)
at java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1402)
at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-58-638.jpg?cb=1557619013" title="58
Asynchronous
def myQuery =
for {
_ <- UIO(println(‚ÄúQuery..." target="_blank">
        58.
      </a>
    58
Asynchronous
def myQuery =
for {
_ &lt;- UIO(println(‚ÄúQuerying!‚Äù))
res &lt;- queryDatabase
} yield res
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-59-638.jpg?cb=1557619013" title="59
Failure!
def myQuery =
for {
_ <- UIO(println(‚ÄúQuerying!..." target="_blank">
        59.
      </a>
    59
Failure!
def myQuery =
for {
_ &lt;- UIO(println(‚ÄúQuerying!‚Äù))
res &lt;- queryDatabase
} yield res
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-60-638.jpg?cb=1557619013" title="60
def myQuery =
for {
_ <- UIO(println(‚ÄúQuerying!‚Äù))
res <..." target="_blank">
        60.
      </a>
    60
def myQuery =
for {
_ &lt;- UIO(println(‚ÄúQuerying!‚Äù))
res &lt;- queryDatabase
} yield res
Fiber:0 ZIO Execution trace:
at myQuery(example.scala:4)
at myQuery(example.scala:3)
Fiber:0 was supposed to continue to:
a future continuation at myQuery(example.scala:5)
Failure!
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-61-638.jpg?cb=1557619013" title="61
def myQuery =
UIO(println(‚ÄúQuerying!‚Äù))
.flatMap(_ =>
qu..." target="_blank">
        61.
      </a>
    61
def myQuery =
UIO(println(‚ÄúQuerying!‚Äù))
.flatMap(_ =&gt;
queryDatabase
.map(res =&gt; res))
Fiber:0 ZIO Execution trace:
at myQuery(example.scala:4)
at myQuery(example.scala:3)
Fiber:0 was supposed to continue to:
a future continuation at myQuery(example.scala:5)
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-62-638.jpg?cb=1557619013" title="62
def myQuery =
UIO(println(‚ÄúQuerying!‚Äù))
.flatMap(_ =>
qu..." target="_blank">
        62.
      </a>
    62
def myQuery =
UIO(println(‚ÄúQuerying!‚Äù))
.flatMap(_ =&gt;
queryDatabase
.map(res =&gt; res))
Fiber:0 ZIO Execution trace:
at myQuery(example.scala:4)
at myQuery(example.scala:3)
Fiber:0 was supposed to continue to:
a future continuation at myQuery(example.scala:5)
The Past
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-63-638.jpg?cb=1557619013" title="63
def myQuery =
UIO(println(‚ÄúQuerying!‚Äù))
.flatMap(_ =>
qu..." target="_blank">
        63.
      </a>
    63
def myQuery =
UIO(println(‚ÄúQuerying!‚Äù))
.flatMap(_ =&gt;
queryDatabase
.map(res =&gt; res))
Fiber:0 ZIO Execution trace:
at myQuery(example.scala:4)
at myQuery(example.scala:3)
Fiber:0 was supposed to continue to:
a future continuation at myQuery(example.scala:5)
The Past
The Future
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-64-638.jpg?cb=1557619013" title="64
def asyncDbCall(sql: SQL): Task[Result]
val selectHumans..." target="_blank">
        64.
      </a>
    64
def asyncDbCall(sql: SQL): Task[Result]
val selectHumans: Task[Result] = ...asyncDbCall(...)...
val selectPets: Task[Result] = ...asyncDbCall(...)...
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-65-638.jpg?cb=1557619013" title="65
def asyncDbCall(sql: SQL): Task[Result]
val selectHumans..." target="_blank">
        65.
      </a>
    65
def asyncDbCall(sql: SQL): Task[Result]
val selectHumans: Task[Result] = ...asyncDbCall(...)...
val selectPets: Task[Result] = ...asyncDbCall(...)...
Fiber:0 ZIO Execution trace:
at asyncDbCall(example.scala:22)
at selectHumans(example.scala:26)
Fiber:0 was supposed to continue to:
a future continuation at selectHumans(example.scala:27)
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-66-638.jpg?cb=1557619013" title="66
def asyncDbCall(sql: SQL): Task[Result]
val selectHumans..." target="_blank">
        66.
      </a>
    66
def asyncDbCall(sql: SQL): Task[Result]
val selectHumans: Task[Result] = ...asyncDbCall(...)...
val selectPets: Task[Result] = ...asyncDbCall(...)...
Fiber:0 ZIO Execution trace:
at asyncDbCall(example.scala:22)
at selectHumans(example.scala:26)
Fiber:0 was supposed to continue to:
a future continuation at selectHumans(example.scala:27)
Gotcha!
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-67-638.jpg?cb=1557619013" title="67
EXECUTION TRACES
def doWork(condition: Boolean) = {
if (..." target="_blank">
        67.
      </a>
    67
EXECUTION TRACES
def doWork(condition: Boolean) = {
if (condition) {
doSideWork()
}
doMainWork()
}
java.lang.Exception: Worker failed!
at example$.doMainWork(example.scala:54)
at example$.doWork(example.scala:50)
at example$$anon$1.run(example.scala:60)
No mention of doSideWork()
PROCEDURAL
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-68-638.jpg?cb=1557619013" title="68
EXECUTION TRACES
def doWork(condition: Boolean) =
for {
..." target="_blank">
        68.
      </a>
    68
EXECUTION TRACES
def doWork(condition: Boolean) =
for {
_ &lt;- IO.when(condition)(doSideWork)
_ &lt;- doMainWork
} yield ()
Fiber:0 ZIO Execution trace:
at example$.doMainWork(example.scala:27)
at example$.doWork(example.scala:23)
at example$.doSideWork(example.scala:26)
The conditional was true!
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-69-638.jpg?cb=1557619013" title="69
CONCURRENT TRACES
uploadUsers uploadPets
uploadTo(target..." target="_blank">
        69.
      </a>
    69
CONCURRENT TRACES
uploadUsers uploadPets
uploadTo(target)
error!
def uploadUsers(users: List[User]): Task[Unit] =
IO.foreachPar_(users.map(toJSON))(uploadTo(dest1))
def uploadPets(pets: List[Pet]): Task[Unit] =
IO.foreachPar_(pets.map(toJSON))(uploadTo(dest2))
def uploadTo(dest: URL)(json: JSON): Task[Unit] = ...
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-70-638.jpg?cb=1557619013" title="70
CONCURRENT TRACES
uploadUsers uploadPets
uploadTo(target..." target="_blank">
        70.
      </a>
    70
CONCURRENT TRACES
uploadUsers uploadPets
uploadTo(target)
error!
java.lang.Exception: Expired credentials
at example$.$anonfun$uploadTo$1(example.scala:28)
Fiber:1 ZIO Execution trace:
at example$.uploadTo(example.scala:28)
Fiber:1 was supposed to continue to: &lt;empty trace&gt;
Fiber:1 was spawned by:
‚ï•
‚ï†‚îÄ Fiber:0 ZIO Execution trace: &lt;empty trace&gt;
‚ïë
‚ïë Fiber:0 was supposed to continue to:
‚ïë example$.uploadUsers(example.scala:21)
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-71-638.jpg?cb=1557619013" title="71
TAGLESS FINAL TRACES
Gain insights into FP libraries
 " target="_blank">
        71.
      </a>
    71
TAGLESS FINAL TRACES
Gain insights into FP libraries
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-72-638.jpg?cb=1557619013" title="72
‚óè Tracing is fast , impact is negligible for real apps
‚óè..." target="_blank">
        72.
      </a>
    72
‚óè Tracing is fast , impact is negligible for real apps
‚óè &gt; 50x faster than Future
‚óè ...even on synthetic benchmarks!
‚óè Impact can be limited by conÔ¨Åg
‚óè Much lower than monad transformers [10x]
‚óè Enabled by default, no Java agents, no ceremony
Disable if tracing is a hot spot
effect.untraced
MADE FOR
PRODUCTION
 
  </li>
  <li>
      <a href="https://image.slidesharecdn.com/errormanagementfuturevszio-190511235539/95/error-management-future-vs-zio-73-638.jpg?cb=1557619013" title="I
73
CONCLUSION
https://github.com/zio
https://gitter.im/zi..." target="_blank">
        73.
      </a>
    I
73
CONCLUSION
https://github.com/zio
https://gitter.im/zio/core
https://zio.dev
@jdegoes
@kaidaxoÔ¨Écial
@shirshovp
 
  </li>

              </ol>
            </div>
      </div>
    </div>

    
  </div>
</div>










      

      

        
    </div></div></div>
    </div>
    <footer>
        <div>created by buildstarted &copy; 2020 <a href="/about">about</a></div>
        <div>Share this page on social media: copy and paste this url https://linksfor.dev/</div>
        <div>If you prefer RSS: <a href="https://linksfor.dev/feed.xml">https://linksfor.dev/feed.xml</a></div>
    </footer>
    
    <script async defer>
        _dna = window._dna || {};
        _dna.siteId = "linksfor.devs";
        _dna.outlink = true;

        (function () {
            let dna = document.createElement('script');
            dna.type = 'text/javascript';
            dna.async = true;
            dna.src = '//dna.buildstarted.com/t.js';
            let s = document.getElementsByTagName('script')[0];
            s.parentNode.insertBefore(dna, s);
        })();
    </script>
    <noscript><img src="//dna.buildstarted.com/g?siteId=linksfor.devs"/></noscript>
</body>
</html>